### DISCUSSION TITLE
Recurring Build Issues with Turborepo: Forced Flags and 'dist' Folder File Deletion

### DISCUSSION THREAD

**Discussion by HamzaAmar** [vercel/turborepo#6840] [Help] [answered]
Hello ,

I hope this message finds you well. I'm currently developing a UI and hooks library using React, managed with Turborepo. Initially, the development process was smooth, but I've encountered a couple of issues that recur:

Forced Flag Usage: For every development script execution, I need to use --force and --no-cache flags. This workaround usually resolves the issue, but it often requires multiple attempts for success.

'dist' Folder File Deletion: When running the development script, it initially works as expected, but eventually, all files within the 'dist' directory are deleted. This issue is critical as my project includes building four interdependent core packages in this order: 'icon', 'utils', 'hooks', and 'core' (the latter relying on all the previous ones). The removal of these files disrupts the dependency chain and complicates the build process.

I would greatly appreciate any insights or solutions for these challenges. Your contributions, in the form of direct assistance or guidance on troubleshooting these issues, would be invaluable. Thank you in advance for your support and willingness to help.

- **Core Package**: Acts as the central hub, integrating `utils`, `hooks`, and `icons` to provide a wide range of features and utilities.

- **Hooks Package**: Offers custom React hooks for enhanced component logic, relying exclusively on the `utils` package for consistent and reliable functionalities.

- **Utils Package**: The independent backbone providing essential utility functions, designed to be used throughout the project without any dependencies.

- **Icons Package**: A standalone collection of icons available for use anywhere in the application, with no external dependencies.
 
 and this the **turbo.json** config
 
```
{
  "$schema": "https://turbo.build/schema.json",
  "pipeline": {
    "build": {
      "outputs": ["dist/**", "storybook-static/**", ".next/**", "!.next/cache/**"]
    },
    "test": {
      "outputs": ["coverage/**"],
      "dependsOn": ["^test"]
    },
    "lint": {
      "cache": false
    },
    "type:check": {
      "cache": false
    },
    "dev": {
      "cache": false,
      "persistent": false,
      "dependsOn": ["^build"]
    },
    "clean": {
      "cache": false
    }
  }
}

```

Here is the repository URL for reference: https://github.com/HamzaAmar/pillar-ui/tree/develop.

Could you, @anthonyshew , @mehulkar, please provide some help? Thank you so much.

**Comment by anthonyshew** [comment]
Could you provide steps to reproduce the undesired behavior with the repo you've linked?

**Reply by HamzaAmar** [reply]
Thank you very much, @anthonyshew, for your prompt response. I encounter an issue when attempting to run the 'dev' script using 'yarn dev.' Is this information sufficient, or would you require further details? Please don't hesitate to let me know if there's anything else you need. I appreciate your assistance and apologize for any inconvenience.

**Reply by anthonyshew** [reply]
Do I need to do anything as far as editing files, a certain, specific string of commands (like starting the dev server, build, then turn it off), or anything else?

**Reply by HamzaAmar** [reply]
Please note, you only need to execute the 'yarn dev' script to proceed. To provide additional context, it's only the development script encountering issues; all other scripts, like the build script, are functioning correctly. Interestingly, when I run the development script individually in each package, it works perfectly. However, difficulties arise when attempting to execute it globally. Your attention and assistance in resolving this matter would be greatly appreciated.

**Reply by anthonyshew** [reply]
I don't receive any errors when I run `yarn dev`.

**Reply by HamzaAmar** [reply]
Could you please repeat it multiple times to see what will happen?

**Reply by anthonyshew** [reply, marked as answer]
Aha, I see errors!

This happens because your `tsup` configurations have `clean: true` in them. Since all of your `dev` tasks are spinning in parallel as persistent tasks, you end up with a race condition amongst your packages. `clean` will wipe out the output folder of the package from disk so, if a dependent package tries to access that output directory while it isn't there, it won't be able to find the modules that it is looking for.

An easy way to make this problem go away is to remove the `clean` flag from your configurations. This comes with the small tradeoff that you could end up with extraneous files in your output directories - but this will rarely happen and you'd find out quickly in your development process if this were to end up being the case.

There are other mitigations for this but this one would be the simplest.

Sidenote: You don't need to `npx tsup` when it is installed in the package's `devDependencies`. You can write `tsup`. Slightly faster, slightly more secure. 👍
